src/ui/public/vis/vis_type.js

```diff
@@ -15,7 +15,12 @@ export default function VisTypeFactory(Private) {
     this.schemas = opts.schemas || new VisTypeSchemas();
     this.params = opts.params || {};
     this.requiresSearch = opts.requiresSearch == null ? true : opts.requiresSearch; // Default to true unless otherwise specified
+    this.listeners = opts.listeners || {};
   }
 
+  VisType.prototype.createRenderbot = function (vis, $el, uiState) {
+    throw new Error('not implemented');
+  };
+
   return VisType;
 };
```

Comment:
> do you maybe know why current implementation is using listeners object with functions which are then attached to events instead of just having vis extend EventEmitter ?

I'm 99% sure that is was just the way that the original implementor wanted to do it. The listeners object though was intended to be a shortcut for situations like [this](https://github.com/elastic/kibana/blob/653fbb5f17d612cdfdb6e93f3c4d5b1170f79875/src/core_plugins/kibana/public/discover/controllers/discover.js#L532-L549) but I totally agree that it would be best to have one way of doing that (and I prefer `vis.on(...)`)

> i would like to understand is the use of all the different event objects 

The `SimpleEmitter` class is a standard event emitter, very similar to the event emitter in node.js. The `Events` class in `ui/events` is an angular compatible version that wraps handlers in promises so that they properly trigger angular digest cycles and bind to the correct angular scope when listeners are added. If the vislib uses the `SimpleEmitter` then any angular consumer has to wrap their event handlers in `$scope.apply()` or something similar to ensure that an angular digest cycle is triggered after running, but if it uses the `ui/events` `Events` class then angular consumers will just be able to consume them like normal.